core/unix/src/TUnixSystem.cxx

```diff
@@ -5148,3 +5275,133 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
 
    return 0;
 }
+
+static void StackTraceFork();
+
+void SetDefaultSignals() {
+   signal(SIGILL, SIG_DFL);
+   signal(SIGSEGV, SIG_DFL);
+   signal(SIGBUS, SIG_DFL);
+}
+
+void TUnixSystem::StackTraceHelperThread()
+{
+   int toParent = fStackTraceHelper.childToParent[1];
+   int fromParent = fStackTraceHelper.parentToChild[0];
+   char buf[2]; buf[1] = '\0';
+   while(true) {
+      int result = SignalSafeRead(fromParent, buf, 1, 5*60);
+      if (result < 0) {
+         SetDefaultSignals();
+         close(toParent);
+         SignalSafeErrWrite("\n\nTraceback helper thread failed to read from parent: ");
+         SignalSafeErrWrite(strerror(-result));
+         SignalSafeErrWrite("\n");
+         ::abort();
+      }
+      if (buf[0] == '1') {
+          SetDefaultSignals();
+          StackTraceFork();
+          SignalSafeWrite(toParent, buf);
+      } else if (buf[0] == '2') {
+          close(toParent);
+          close(fromParent);
+          toParent = fStackTraceHelper.childToParent[1];
+          fromParent = fStackTraceHelper.parentToChild[0];
+      } else if (buf[0] == '3') {
+          break;
+      } else {
+          SetDefaultSignals();
+          close(toParent);
+          SignalSafeErrWrite("\n\nTraceback helper thread got unknown command from parent: ");
+          SignalSafeErrWrite(buf);
+          SignalSafeErrWrite("\n");
+          ::abort();
+      }
+   }
+}
+
+void TUnixSystem::StackTraceFromThread()
+{
+   int result = SignalSafeWrite(fStackTraceHelper.parentToChild[1], "1"); 
+   if (result < 0) {
+      SignalSafeErrWrite("\n\nAttempt to request stacktrace failed: ");
+      SignalSafeErrWrite(strerror(-result));
+      SignalSafeErrWrite("\n");
+      return;
+    }
+    char buf[2]; buf[1] = '\0';
+    if ((result = SignalSafeRead(fStackTraceHelper.childToParent[0], buf, 1)) < 0) {
+       SignalSafeErrWrite("\n\nWaiting for stacktrace completion failed: ");
+       SignalSafeErrWrite(strerror(-result));
+       SignalSafeErrWrite("\n");
+       return;
+    }
+}
+
+void StackTraceFork()
+{
+   char childStack[4*1024];
+   char *childStackPtr = childStack + 4*1024;
```

Comment:
if a constant is repeated, it ought to be declare as a constexpr to be used consistently.